### PR TITLE
[RISCV] Insert a freeze before converting select to AND/OR.

### LINKED ISSUE #84200: [RISCV] Miscompile at -O3
Reduced test case: https://godbolt.org/z/1sxbE8eco
```
#include <stdint.h>
#include <stdio.h>
int32_t a, f;
uint16_t e = 10;
static uint8_t f1(uint8_t c, uint8_t d) { return d == 0 ? 0 : c % d; }
static uint32_t f2(uint32_t c, uint32_t d) { return d == 0 ? c : 0; }
static int32_t *h(int32_t l) {
  int32_t *n = &a;
  return n;
}
static uint64_t g(int16_t k) {
  uint16_t m[] = {0, 3};
  int32_t *j = h(0);
  *j = f1(f2(1 && m[1], 0), e) == 0;
  return m[f];
}
int main() {
  g(f);
  printf("%d\n", a);
}
```

```
> clang -O0 test.c && ./a.out
0
> clang -O3 test.c --target=riscv64-linux-gnu
> qemu-riscv64 -L /usr/riscv64-linux-gnu/ ./a.out
-10
```

Clang version: 6e27dd47e1fc17cf661b568901fe7392ed1b8f2b

### PR DESCRIPTION
**Pull request by topperc**
Select blocks poison, but AND/OR do not. We need to insert a freeze
to block poison propagation.

This creates suboptimal codegen which I will try to fix with other
patches. We should prioritize the correctness fix.

Fixes #84200.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -7274,6 +7274,7 @@ static SDValue combineSelectToBinOp(SDNode *N, SelectionDAG &DAG,
     const APInt &FalseVal = FalseV->getAsAPIntVal();
     if (~TrueVal == FalseVal) {
       SDValue Neg = DAG.getNegative(CondV, DL, VT);
+      FalseV = DAG.getFreeze(FalseV);
```
Comment (nikic) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7277]: This freeze looks unnecessary. If x is poison so is ~x and they're constants here to boot.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -7289,14 +7290,14 @@ static SDValue combineSelectToBinOp(SDNode *N, SelectionDAG &DAG,
     // (select x, x, y) -> x | y
     // (select !x, x, y) -> x & y
     if (std::optional<bool> MatchResult = matchSetCC(LHS, RHS, CC, TrueV)) {
-      return DAG.getNode(*MatchResult ? ISD::OR : ISD::AND, DL, VT, TrueV,
-                         FalseV);
+      return DAG.getNode(*MatchResult ? ISD::OR : ISD::AND, DL, VT,
+                         DAG.getFreeze(TrueV), DAG.getFreeze(FalseV));
```
Comment (nikic) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7294]: You don't need to freeze x, as it's part of the select condition.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -7289,14 +7290,14 @@ static SDValue combineSelectToBinOp(SDNode *N, SelectionDAG &DAG,
...
     }
     // (select x, y, x) -> x & y
     // (select !x, y, x) -> x | y
     if (std::optional<bool> MatchResult = matchSetCC(LHS, RHS, CC, FalseV)) {
-      return DAG.getNode(*MatchResult ? ISD::AND : ISD::OR, DL, VT, TrueV,
-                         FalseV);
+      return DAG.getNode(*MatchResult ? ISD::AND : ISD::OR, DL, VT,
+                         DAG.getFreeze(TrueV), DAG.getFreeze(FalseV));
```
Comment (nikic) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7300]: Same here.

**Comment by nikic** [comment]
Keep in mind that we still do the logical and/or -> bitwise and/or fold in generic DAGCombine. That's the really big offender in terms of poison safety in SDAG.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (nikic) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: You do still need the freeze on y (FalseV), just not on x (TrueV).

File: llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64zbb.ll [hunk 4]
```diff
@@ -481,12 +481,10 @@ define signext i32 @ffs_i32(i32 signext %a) nounwind {
 ; RV64I-NEXT:    addi a1, a1, %lo(.LCPI9_0)
 ; RV64I-NEXT:    add a0, a1, a0
 ; RV64I-NEXT:    lbu a0, 0(a0)
-; RV64I-NEXT:    addi a0, a0, 1
+; RV64I-NEXT:    addiw a0, a0, 1
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/rv64-legal-i32/rv64zbb.ll:484]: It seems to be a happy surprise? we removed a zero extend here.

**Comment by dtcxzyw** [comment]
Should we insert a `freeze` node in the following code as well?

https://github.com/llvm/llvm-project/blob/91808c845fd6f0624525b6d6348b2c284628ce58/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L7392-L7410

**Review by nikic — COMMENTED** [review]
This looks fine to me, as long as RISCV folks are fine with having these regressions in the meantime. The alternative would be to land mitigations first.

**Comment by dtcxzyw** [comment]
There is no regression in the LLVM test-suite :)

**Comment by wangpc-pp** [comment]
> There is no regression in the LLVM test-suite :)

Try `-march=rv32gc`? It seems there are a lot of regressions.

**Comment by dtcxzyw** [comment]
> > There is no regression in the LLVM test-suite :)
> 
> Try `-march=rv32gc`? It seems there are a lot of regressions.

https://github.com/dtcxzyw/llvm-ci/pull/1106#issuecomment-1983998365

**Comment by preames** [comment]
> This looks fine to me, as long as RISCV folks are fine with having these regressions in the meantime. The alternative would be to land mitigations first.

LGTM specifically on the taking the regressions for correctness point.  I'm assuming Nikita's correct on the poison propagation rules.  I glanced at the code, didn't see any obvious errors, but this stuff makes my head hurt.

**Review by preames — APPROVED** [review]

**Comment by bjope** [comment]
I think we might have hit a similar problem in our testing, but involving foldBoolSelectToLogic in generic DAGCombiner.

The IR input to llc involved
```
  %x = load i16, ptr @v_473, align 1, !tbaa !11
  %cmp1 = icmp sgt i16 %x, 0
  %sub = sub nsw i16 32767, %x
  %cmp2 = icmp sgt i16 %x, %sub
  %y = select i1 %cmp1, i1 %cmp2, i1 false
```
Resulting in a DAG with:
```
     t6: i1 = setcc t4, Constant:i16<0>, setgt:ch
        t8: i16 = sub nsw Constant:i16<32767>, t4
      t9: i1 = setcc t4, t8, setgt:ch
    t11: i1 = select t6, t9, Constant:i1<0>
```
that is combined into
```
    t6: i1 = setcc t4, Constant:i16<0>, setgt:ch
        t8: i16 = sub nsw Constant:i16<32767>, t4
      t9: i1 = setcc t4, t8, setgt:ch
    t23: i1 = and t6, t9 
```
If t6 is false the result should be zero. But after folding to the and the result would become t9 instead.
But when t6 is false, it means that t4<=0. And for t4<0 then t8 is poison, so t9 could be poisonous.

**Comment by bjope** [comment]
> If t6 is false the result should be zero. But after folding to the and the result would become t9 instead.

Hm. Something wrong in my reasoning here. If t6 is false the and will return false. But maybe the rewrite into and is wrong anyway?

**Comment by bjope** [comment]
I wrote https://github.com/llvm/llvm-project/issues/84653 about the problems I've seen with foldBoolSelectToLogic rewriting SELECT into AND without checking isGuaranteedNotToBeUndefOrPoison on the operands.

**Pull request merged**